### ISSUE TITLE
Unary operator trunc gives incorrect results for negative numbers

### ISSUE THREAD

**Issue by Beechside** [supercollider/supercollider#2378]
The unary operator trunc should always round towards zero.

-3.3.trunc should give -3 but SC gives -4
-3.3.trunc(0.25) should give -3.25 but SC gives -3.5

I know of no language where trunc always rounds towards negative infinity so believe this to be a bug in SC.

I realise that I can use -3.3.trunc(-1) and -3.3.trunc(-0.25) to get the correct results but that seems to be going against the "standard" definition of trunc.

**Comment by Sciss** [comment]
SuperCollider defines many mathematically operations in a "non-standard" way. Given that trunc has worked this way since 2003 or so, I don't think it's warranted to change its operation. Apart from this general reservation, I'm also not aware that "trunc" is defined anywhere as a standard mathematical operation.

Trunc is defined as follows (my copy in Scala):

```scala
def trunc (a: Float, b: Float): Float = if (b == 0) a else (math.floor(a / b) * b).toFloat
```

This is a very clear and unambiguous definition.

**Comment by Beechside** [comment]
I hope you don't mind my saying that your example is a clear implementation, rather than a clear definition ;o)

I can understand the desire to protect existing code and will support if that is the broad consensus.

Nonetheless, the act of truncation is to remove digits, not to round or to floor.  Removing digits always tends towards zero.  As for a "standard" definition, you would find no universal algorithmic consensus anywhere from from wiki to Knuth but truncation is a very unambiguous process which I believe to have a clear mathematical parallel.

I appreciate your speedy reply - thank you from a github newcomer, even if I am an SC veteran...

**Comment by vivid-synth** [comment]
Other languages do agree with @Beechside's description:

Haskell:
```
> truncate (-3.3)
-3
```

Ruby:
```
irb(main):003:0> -3.3.truncate
=> -3
```

Python:
```
>>> import math
>>> math.trunc(-3.3)
-3
```

Javascript:
```
Math.trunc(-3.3)
-3
```

I'm maybe not the best advocate because I don't use sclang as much as scsynth, but I'd be open to the idea of "fixing" how ```trunc``` works if there isn't much code that uses it with negative numbers, and if everyone is notified that it's changed. Not sure it's worth it on balance, but it's worth discussing.

**Comment by vivid-synth** [comment]
An argument in favor of the "fix" is that ```trunc``` has such a standard understood meaning that newcomers (and not-so-newcomers) may reach for it without looking up its definition, because it's "obvious" what it does.

At the very least, we should prominently note this behavior in the docs, although Nick Coghlan (a core Python dev) has a thing or two to say about that:

> There's a great saying in the usability world: "You can't document
> your way out of a usability problem". What it means is that if all the
> affordances of your application (or programming language!) push users
> towards a particular logical conclusion [...], having a caveat in your documentation isn't
> going to help, because people aren't even going to think to ask the
> question.

 (https://mail.python.org/pipermail/python-ideas/2014-March/026647.html)

**Comment by telephon** [comment]
I see what you mean. Here, SuperCollider simply inherits `trunc` from the C implementation:

http://www.cplusplus.com/reference/cmath/trunc/?kw=trunc


```
inline float32 sc_trunc(float32 x)
{
	return std::trunc(x);
}
```

Use `round` if you want the math op you describe.

**Comment by Beechside** [comment]
@telephon "Use `round` if you want the math op you describe."

Were it that easy...

Firstly, `-3.7.round` would give -4 (just as incorrect as `-3.7.trunc`) so I would need to have different code for decimal parts above and below 0.5

Secondly, I don't really want to have one method for positive numbers and another for negative.  I want to use variables in a single, simple statement that gives predictable and "correct" results.

I still believe `trunc` should truncate, i.e. round towards zero, not towards the nearest x.

**Comment by Beechside** [comment]
Just one more thing:  I first raised this in 2013 in the SuperCollider mailing list:

http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/Integer-division-td7600460.html#a7600490

At the time, Tim Blechman and HJH thought it a bug and were going to report it, though I haven't found it in this repository.  I've just come back to some quantisation code I had written and found the problem again, hence my raising it here. There may be some useful correspondence if anyone can find the log.

**Comment by telephon** [comment]
@Beechside sorry, all this was maybe a little too quick! 

I was actually looking for a label "works-according-to-spec" but didn't find it :)

**Comment by telephon** [comment]
You are right about `round`. We could add a `trunc2` op, which would implement the other behavior. The `2`would make sense in the sense of symmetry, I think.

**Comment by Beechside** [comment]
@telephon Thank you - I appreciate your comments.

However, I'm confused by your reference to the cplusplus site. The examples cited there produce the results I was expecting (-3.8.trunc becomes -3, unlike SC, which returns -4).  The page specifically says "Round x towards zero"

Ditto the isocpp.org site.  So, I can't see a way to label SC's behaviour: "works-according-to-spec".

It seems to me that the choice is "correcting" the code to conform to normal expectation but potentially breaking existing user code or introducing a new op.  At a personal level, I appreciate the deprecation and conformance (eg cross-platform GUI) that has taken place in recent releases.  It seems fundamentally wrong to introduce a new method to fudge a fault (if I may call it that).

For Integers, `floor` mimics the behaviour of the current `trunc` but doesn't work for Float.  May I propose the following?

`trunc` to be recoded to behave as it should - rounding towards zero
`floor` and `ceil` are converted to binary operators to mimic behaviour of round _und seine Freunde_, allowing specification of the target result.

OK, it's more work and is not exactly high priority but I believe it follows the principle of Occam's Razor.  It would also mean warning users when released but is certainly no worse than the recent deprecations, which I warmly welcome.

Apologies if I appear too strong or opinionated.  Please put any offence down to my novice status, not to intended rudeness...

**Comment by danstowell** [comment]
FWIW I'd suggest that trunc is pretty niche, and so fixing it is OK (I wouldn't expect it to break too much code: note that we're expecting its behaviour on positive values to remain the same).

**Comment by telephon** [comment]
@Beechside It turns out that the description on the page is really different, probably because sc uses `std:trunc`? Very bad.

@danstowell `trunc` has been around as it is for really long, and I simply don't know if those "niches" it has been used for will cause very hard to find "niche" errors. Don't you think that it is better to just add `trunc2`?

**Comment by danstowell** [comment]
Yes, I do see that. But adding `trunc2` (without altering `trunc`) would be beside the point

**Comment by Beechside** [comment]
I'm not sure I can add much more, since I'm not an experienced member of the dev community and don't know how such changes are approved/prioritised.  

I was using `trunc` to weight random processes around a target note in voice leading and was curious why it was not working for notes below my tonal centre.  I wasted time in diagnosing SuperCollider - something we shouldn't ask of our users.

Given that there is a lot more code to be written with SuperCollider than has so far been written, I humbly vote for getting it right.  This means no deprecation but immediate correction and notifying the user community of the change, which I too think is pretty esoteric.  

I would also much prefer binary `floor` and `ceil`, giving the full range of quantisation tools.

**Comment by telephon** [comment]
I agree that it would be a better life with a fixed trunc. I'l like to collect some other voices and let it sink a little.

I'm not sure still how much it would break, or if we could easily tell. See e.g.
http://doc.sccode.org/Classes/VBAP.html  But well, let's be optimistic.

With regards to the binops for floor and ceil: so `1.3.ceil(0.5)` would actually floor? These functions have a clear mathematical definition (for once!) and maybe it is better to add a `roundDown` method (effectively `roundUp() - 1`).

**Comment by gusano** [comment]
> I'l like to collect some other voices and let it sink a little

@telephon my vote goes up to fixing `trunc` without adding a new method.
It has used a wrong behaviour for much too long so having a proper `trunc` method in "modern" sc is worth potentially breaking (a few) code :)

**Comment by telephon** [comment]
that said, it is still unclear to me why our `sc_trunc` primitive does what it doesn't do. (`return std::trunc(x);`).

**Comment by vivid-synth** [comment]
@telephon it looks like you were looking at UnaryOps -- but in fact the behavior is defined for BinaryOps, which is defined like this:

```
inline double sc_trunc(double x, double quant)
{
        return quant==0. ? x : sc_floor(x/quant) * quant;
}
```

**Comment by vivid-synth** [comment]
(E.g. in SCClassLibrary/Common/Math/SimpleNumber.sc:)

```
trunc { arg aNumber=1.0, adverb; _Trunc; ^aNumber.performBinaryOpOnSimpleNumber('trunc', this, adverb) }
```

**Comment by Beechside** [comment]
@telephon I'm happy to raise another ticket if this is better but, no, `1.3.ceil(0.5)` would not floor.  In the current implementation, `ceil` raises to the next higher integer.  But, were it a binary operator, it would raise to the next `arg` so, in this case 1.5.  For clarity:

`1.3.ceil = 2`
`-1.7.ceil = -1`
`1.3.ceil(0.5) = 1.5`
`-1.7.ceil(0.5) = -1.5`

`1.7.floor = 1`
`-1.3.floor = -2`
`1.7.floor(0.5) = 1.5`
`-1.3.floor(0.5) = -1.5`

So, would the binOps actually be used?  They are, essentially, quantisation tools, whether for pitch or rhythm so, I can see a use for them.  As you've certainly hit in the past, adding 0.5 and rounding is not a valid substitute (eg when the object is an integer value).

Are these the same as roundUp and the proposed roundDown?  I can't see any difference so I think you are right that roundDown would be a useful, and nicely symmetrical, addition.

**Comment by telephon** [comment]
@Beechside thank you for the explanation, I see now how it is meant, and that rounding and flooring can be different, yet analogous.

All these methods have symmetric counterparts. This is because "down" may correspond to "towards zero" or "towards -inf".

Here is the "full" collection:
```
ceil - floor (trunc = floor(1))
roundDown roundUp

```

I'd propose to make this a large and systematic change, not a small fix.

We should simply have both "logics" fully implemented. I would find it logical to have the "2" on the end for the symmetric variants (analogous to e.g. `clip2`), but whatever seems pleasing.

**Comment by Beechside** [comment]
I am a little confused by `trunc = floor(1)` for two reasons (possibly my misunderstanding of your syntax):

1) `floor` always rounds down so `-3.2.floor` would result in -4 but `-3.2.trunc` should result in -3.  Truncation should always round towards zero, since it reduces the magnitude of the object but doesn't change its sign (except truncating a negative number giving a result of zero).

2) `trunc` does not always give an integer result, whereas `floor(1)` would.  The current implementation of `trunc` already allows for a rounding factor and this should remain.

As @danstowell says,`trunc` works correctly for all positive SimpleNumbers.

By the way, I think that creating `trunc2`, which always rounds to negative infinity would be a mistake.  It is not truncation in any mathematical sense and the function is catered for by the proposed `roundDown` method.  To call it `trunc2` would be misleading.  `roundDown`, on the other hand, is perfectly named ;o)

Thank you all for your ongoing contribution.  It is hugely valuable.

**Comment by vivid-synth** [comment]
Note this doesn't just affect SimpleNumber - it also affects UGen signals. Still, a quick look at a couple hundred sccode pieces only shows one use of '.trunc', and it's only given positive numbers.

**Comment by Sciss** [comment]
On 06/10/16 15:48, vivid-synth wrote:
> Note this doesn't just affect SimpleNumber - it also affects UGen
> signals. Still, a quick look at a couple hundred sccode pieces only
> shows one use of '.trunc', and it's only given positive numbers.

That's why I am also not so reluctant at fixing this as I would have
usually been (because I think the value of backwards compatibility is
underestimated).

If we saw that the current behaviour was relied on, the story would be
different. But probably you are right, and the intentional use of the
behaviour for negative numbers is very rare.

**Comment by snappizz** [comment]
FWIW, I agree with Hanns. Changing a core math function, no matter how rarely used, is disruptive.

The server-side effects, aside from the obvious change to BinaryOpUGen, would affect OSTrunc4 and OSTrunc8 in sc3-plugins.

**Comment by snappizz** [comment]
(Deleted my last message since I wasn't reading carefully enough)

The server-side effects, aside from the obvious change to BinaryOpUGen, would affect OSTrunc4 and OSTrunc8 in sc3-plugins.

**Comment by telephon** [comment]
… and it would affect any code that uses `trunc` on a ugen.

**Comment by telephon** [comment]
>By the way, I think that creating `trunc2`, which always rounds to negative infinity would be a mistake.

I was suggesting that `trunc2` and all the corresponding other methods would act symmetrically around zero. This would leave `trunc` in place. It is a common way in sclang to do this.

**Comment by Beechside** [comment]
While there is a functional parallel between `clip` and `clip2` and similar pairings, there is no mathematical link been a correct `trunc` and a `trunc2`.  By definition, truncation removes digits, thereby reducing the magnitude or absolute value of a number.  We may then choose to round the result but it's magnitude can never be larger than the original value.  Truncation can never increase absolute value, hence my saying that creating a `trunc2` (round to negative infinity) function is confusing and arguably wrong.  Any such function should not be called truncation, though it might be given a different name.

I get the impression from others that `trunc` should behave as for other languages, even if there is a remote chance of backwards incompatibility.  Not knowing how such changes are approved, I'll stand back and watch...

**Comment by danstowell** [comment]
@Beechside I think you misread @telephon slightly: he was suggesting that `trunc` should remain unchanged and `trunc2` should implement what you think of as trunc behaviour. I disagree with him on that. The advantage of his suggestion is backward compatiblity since `trunc` would remain unchanged. But to me that's not fixing the problem. And most opinion here so far seems to be that the API change is a small price to pay.

**Comment by Beechside** [comment]
@danstowell Ah.  I was relying too much on @telephon saying this: "I agree that it would be a better life with a fixed `trunc`."  I now see what you mean.

Having mentally renamed `trunc` as `roundIn`, I think the current `trunc` function is better named roundDown, a method suggested by @telephon.  We should perhaps offer a quintet of `round` methods:

- `round` – strict mathematical rounding.  We cannot predict direction in advance
- `roundUp`– all results tend to +inf
- `roundDown`– all results tend to -inf (the current `trunc` function)
- `roundIn` – all results tend to 0 (with an alias of `trunc`)
- `roundOut` – all results tend away from 0 (not currently implemented)

Currently, `ceil` is a unary version of roundUp and `floor` is a unary version of `roundDown`.  I think little harm can come from pointing both at their binop cousins.  The code base is simplified and both extend their function.

**Comment by danstowell** [comment]
I'm not interested in those new names. Much more sensible to me are the names ceil, floor, trunc and round, whether unary or binary.

**Comment by Beechside** [comment]
Noted.  I was shooting for a full house ;o)

There is already a binary version of `ceil`, called `roundUp` and the binary version of `floor` is the currently implemented `trunc`, which needs to go somewhere so `roundDown` seems as good as anything.

`roundOut` seems to complete the set but, in the absence of anyone asking for it before, it may be superfluous.  These basic transformations can always be achieved in other ways and are, I suppose, conveniences...

**Comment by HotwheelsSisyphus** [comment]
I ran into this again recently and am just going to necro-bump it here. It does seem like a pretty "important" thing (in the sense that this is something one relies on to work as expected  if one knows a bit of [math](https://en.wikipedia.org/wiki/Truncation) and probably any [programming language](https://cplusplus.com/reference/cmath/trunc/?kw=trunc) that isn't sc). And it seems this was more forgotten about than decided on, so here goes:

**At minimum, I suggest that I document this nonstandard behavior for the time being**
(because this is hopefully easier to agree on) although I strongly agree with the thing said ten years ago:

> An argument in favor of the "fix" is that `trunc` has such a standard understood meaning that newcomers (and not-so-newcomers) may reach for it without looking up its definition, because it's "obvious" what it does.
> 
> At the very least, we should prominently note this behavior in the docs, although Nick Coghlan (a core Python dev) has a thing or two to say about that:
> 
> > There's a great saying in the usability world: "You can't document
> > your way out of a usability problem". What it means is that if all the
> > affordances of your application (or programming language!) push users
> > towards a particular logical conclusion [...], having a caveat in your documentation isn't
> > going to help, because people aren't even going to think to ask the
> > question.
> 
> (https://mail.python.org/pipermail/python-ideas/2014-March/026647.html

-----------------------------------
In hopes that this might help towards a consensus, I'm quoting to recall the (most recent) positions of the folks in this discussion. 

@Beechside raised the issue, so presumably in favor of API change

@vivid-synth in favor, as per the above quote about how "You can't document your way out of a usability problem" and the numerous examples from other languages

@danstowell 
> @telephon was suggesting that `trunc` should remain unchanged and `trunc2` should implement what you think of as trunc behaviour. I disagree with him on that. The advantage of his suggestion is backward compatiblity since `trunc` would remain unchanged. But to me that's not fixing the problem. And most opinion here so far seems to be that the API change is a small price to pay.

@gusano
> my vote goes up to fixing `trunc` without adding a new method. It has used a wrong behaviour for much too long so having a proper `trunc` method in "modern" sc is worth potentially breaking (a few) code :)

@jamshark70 @timblechmann in a mailing list anno 2013 (according to @Beechside)
> thought it a bug and were going to report it, though I haven't found it in this repository. I've just come back to some quantisation code I had written and found the problem again, hence my raising it here. There may be some useful correspondence if anyone can find the log.

@Sciss (originally more against API change)
> On 06/10/16 15:48, vivid-synth wrote:
> 
> > Note this doesn't just affect SimpleNumber - it also affects UGen
> > signals. Still, a quick look at a couple hundred sccode pieces only
> > shows one use of '.trunc', and it's only given positive numbers.
> 
> That's why I am also not so reluctant at fixing this as I would have usually been (because I think the value of backwards compatibility is underestimated).
> 
> If we saw that the current behaviour was relied on, the story would be different. But probably you are right, and the intentional use of the behaviour for negative numbers is very rare.

@telephon (mostly against API change for reasons summarized in the Dan Stowell quote)
> I'l like to collect some other voices and let it sink a little

@nhthn and @rd--  did not pronounce an opinion.

-------------------------------------------------------------------------------------------

And here's a short summary about what the change would affect (repeating stuff already quoted; and I have not verified this):

@nathan:
> The server-side effects, aside from the obvious change to BinaryOpUGen, would affect OSTrunc4 and OSTrunc8 in sc3-plugins.

@telephon:
> … and it would affect any code that uses trunc on a ugen. [because most ugens are bipolar]

@vivid-synth 
> Still, a quick look at a couple hundred sccode pieces only shows one use of '.trunc', and it's only given positive numbers.


Also worth recalling: This is actually about the _binary_ op sc_trunc, not the unary op sc_trunc which already forwards to  std::trunc.

**Comment by jamshark70** [comment]
If I search my entire local SC support directory, I find 42 uses of `trunc` excluding HelpSource to check for the possibility of negatives.

I wonder what other people in this thread will find.

Just to assess "And most opinion here so far seems to be that the API change is a small price to pay" -- how small is the price actually.

**Comment by telephon** [comment]
I agree with @Sciss first comment: while nonstandard, it is a perfectly reasonable way to understand truncation. Besides, changing trunc is definitely not worth the portential trouble. Let's add trunc2.

**Comment by Beechside** [comment]
Last comments from me as OP before I unsub from this thread.

I'm not sure that looking at library usage of trunc will help.  I have a lot of old sclang but it is no longer current.  Harder to do, but a better metric were it possible, would be to ask what users expected.  I asked my students what they expected trunc to do and all without exception said "drop the decimal component".

Breaking changes are common when updating software.  It's an honest admission that either something was wrong or that something can be improved and is worth the impact.  I'm currently doing a lot with Python and C#.  Both have histories of breaking changes and result in, arguably, greater global impact than sclang.

As I said once before, there is far more code waiting to be written in sclang than has been written so far.  My vote would still be to fix what is an error.  Mark it breaking and live with it. 

Creating trunc2 puts a cognitive load on users that is, frankly, wrong.

As always, written in a spirit of open source collab.  There are no kings :o)

**Comment by HotwheelsSisyphus** [comment]
> I agree with [@Sciss](https://github.com/Sciss) first comment: while nonstandard, it is a perfectly reasonable way to understand truncation

I beg to differ here. I can get on board with the backward compatibility argument if that should turn out to be the consensus, but not with this one.
That definition is simply **not** truncation, it's more a kind of quantization, or simply the floor function if b is 1.  Also not clear where @sciss had got it from. 
Yes, when the domain is positive, the floor function happens to coincide with truncation to integers but that doesn't mean that's "a reasonable way to understand truncation". There's a reason why people make the distinction. There's also a reason why it's called "truncation", as in you simply "chop off" n-ary digits below a given precision. 
Obviously one can define any function as anything, and that's ok, but at the end of the day it's a question about communication: What do other people expect from the thing that I call trunc? 

I tend to also  agree with @Beechside's point about:
>  there is far more code waiting to be written in sclang than has been written so far.

-------------------------------------

If we leave trunc unchanged (or even, unchanged "for the time being", e.g. the next 10 years), then what remains to be decided on is whether to put a replacement function in there and what to call it. 

@telephon suggested **trunc2**, by analogy to clip2 etc. 
other options I can think of:
* **truncate** (I don't think it's necessarily _more_ confusing to have both a .truncate and a .trunc method, because by the time you have to look up .trunc in the sc docs, you're _already_ confused.)
* ~~.**fix** (that's what it's called in matlab, for whatever reason; )~~
* .**truncFixed**, **truncCorrected** (or something like that)...
* ~~.**mathTrunc**~~, **truncMath** ...
* **truncToward0** 

it should probably start with "trunc" so it shows up in people's autocomplete (hence the strikethrough)
by the same token, .**truncCorrected** (or truncFixed, or similar) is probably good because it will show up in autocomplete and also implies that there is something _not_ corrected/fixed about the regular trunc. (Similary, if it was called truncToward0 it would imply, correctly, that the other trunc does not truncate toward 0).

I think in this case it's not a problem if the selector is longer, because realistically it's not something that anyone will be typing multiple times every day. Might even be easier to notice this way.

(
Another possibility: don't define an additional method, but include an alternative function in the note/warning about the regular (broken) trunc method, e.g.:
```
// For reasons of backward compatibility, SuperCollider's trunc method rounds toward negative infinity everywhere. For the standard behavior of trunc (round toward 0 everywhere), you can use the following function:

t = { |x, prec=1| abs(x).div(prec) * prec * x.sign }
```
Of course that may not be efficient and people will _not_ see that in their autocomplete.
)

**Comment by cdbzb** [comment]
I dislike trunc2 as there is no hint what trunc and trunc2 actually do. There are many examples of this kind of naming in SC and they are unfortunate imo.

The behavior of trunc is unexpected and it should be changed or deprecated in favor of a new method.

We run into this exact discussion all the time - OT but we should perhaps settle on a policy!

I vote for fixing it and providing truncOld as a drop-in for legacy code. I just hate the idea that new users will reach for trunc and have to read docs before picking which method to use and also read code and not understand it correctly if they were not already aware of this quirk!

Of the options preserving backwards compatibility I like 'truncate' ...in which case we should soft-deprecate 'trunc'!

**Comment by dyfer** [comment]
IMO we have two options to address this, along the lines of what @cdbzb wrote:
- option 1: break some code, clean up behavior; maintain shorter method
  - fix `trunc` to do what it's supposed to do; _document the change_
  - add `truncOld` with old behavior
- option 2: non-breaking with deprecation, but adds more methods and more typing
  - deprecate `trunc` (display warning when used)
  - implement `truncate` with the correct behavior
  - add `truncOld` with old behavior
- EDIT: option 3: not deprecate but document the current trunc and add a new trunc2 that acts as expected (proposed by @telephon)

EDIT: maybe we can have an informal vote with emojis?
option 1: 🚀
option 2: 😄
neither/something else: 👀

**Comment by telephon** [comment]
>We run into this exact discussion all the time - OT but we should perhaps settle on a policy!

Yes, there has been a practiced policy about this, not to change implementation in such cases, also not to deprecate if not absolutely necessary. You suggest to change this policy. I would prefer to keep it.

But I am also not annoyed by clip ves clip2 etc. so it is hard for me to follow the argument. I am sorry if I am being stubburn here.

**Comment by telephon** [comment]
**option 3:** not deprecate but document the current trunc and add a new trunc2 that acts as expected. 

For reference, an analogous case is `pow` which also behaves differently, albeit for somewhat better reasons. It has been often complained about nevertheless.

Again, the trunc behaviour is logical, there are simply two ways to do truncation:
-  downwards towards absolutely smaller numbers (asymmetric) 
-  towards the neutral element zero (symmetric). 

We often denote symmetric operations with a 2, so this is lucky. E.g. clip2 is symmetric. So analogously, trunc2 would be symmetric.

**Comment by HotwheelsSisyphus** [comment]
> a practiced policy

It would be good to at least make this practiced policy explicit at a coming dev meeting; and in a separate step, one could then discuss whether to change that policy or not (so it's clear what actually is being discussed and what kinds of arguments are relevant). I imagine that this kind of discussion about general policy might benefit from being decoupled from the discussion about trunc in particular.

I was simply not aware of a policy in this regards. To a relative newcomer like myself, situations like the one concerning trunc do not look like they're informed by policy at all, also because the nonstandard behavior is not even documented. It just looks like nothing has ever been decided.

**Comment by dyfer** [comment]
> Yes, there has been a practiced policy about this, not to change implementation in such cases, also not to deprecate if not absolutely necessary.

Correct me if I'm wrong, but I don't think this has been explicitly written down as a policy. It would be good to do that. Moss has written this down in a [project-wide todo](https://github.com/supercollider/supercollider/wiki/Project-tasks-organization#issues-by-area) a while back (though that document is somewhat outdated).

We do have a minimal policy for [deprecating classes](https://github.com/supercollider/supercollider/wiki/git-workflow-and-guidelines#deprecation), but not methods, it seems.


> **option 3:** not deprecate but document the current trunc and add a new trunc2 that acts as expected.
> ...
> Again, the trunc behaviour is logical, there are simply two ways to do truncation:

I'm on board with that option too.

**Comment by cdbzb** [comment]
> Again, the trunc behaviour is logical, there are simply two ways to do truncation:

Is there a single example in any other language of truncation being asymmetrical? I cannot find one.

https://en.wikipedia.org/wiki/Truncation - "truncation always rounds toward zero"

**Comment by JordanHendersonMusic** [comment]
> Again, the trunc behaviour is logical, there are simply two ways to do truncation:

If trunc does floor, trunc2 decreases the absolute value, then that implies trunc3 should do ceil!

I think trunc2 is the best available compromise given we lack the language level features to do this properly. Not pretty, nor nice for users, but our hands are tied.

**Comment by telephon** [comment]
About making the implicit explicit: It would be useful for everybody involved to have a short text that unfolds the dimensions of such decisions and how we have been solving them. This would make them open to discussion in principle, independently of the particular case. 

It would help navigate the aesthetic, social and computational aspects of trade-offs like this, we would then perhaps discuss more about where to place a specific case on such a map, keeping in mind all of the trade offs.

It would be fun to think this through, I could come up with a draft for the next dev meeting.

**Comment by jamshark70** [comment]
The backward compatibility issue is less about active users who are likely to be reading changelogs, but rather for that composer somewhere who wrote a SC piece in 2017 and never really used SC very much since, and now the piece is being programmed on a concert again, so they download SC 3.14 and suddenly some details that were minor to us aren't working like they used to. They're not going to go through 8 years of changelogs.

Pamela Z admitted in her ICMC 2023 keynote that she's still on Max 5 because her live patch broke in Max 6.

I don't think changing trunc to be standard would affect *me*, but it isn't only about me. I'm *really* sympathetic to the idea of changing it actually! But, SC3 is mature. We should really be working toward SC4. It's time.

**Comment by JordanHendersonMusic** [comment]
> They're not going to go through 8 years of changelogs.

This!

> We should really be working toward SC4. It's time.

I'm currently putting together a prototype and my intention is to try and get community support and some funding. It will have versioned libraries, which will solve this issue. 

Actually, not having to have these complex and time consuming discussions every month about backwards compatibility is one of the key features.

**Comment by capital-G** [comment]
> Pamela Z admitted in her ICMC 2023 keynote that she's still on Max 5 because her live patch broke in Max 6.

Well, thats how you should do it and that is also best practice in the software industry. If you write a piece in Max 5, use Max 5 to run it.
If you change your runtime, all things can go wrong (e.g. even packages such as pytorch have problems updating to a new python runtime and break). Thats what patch versions are for.

If the runtime is not supported by your OS anymore, you have to virtualize it (but this is a bad idea for realtime related environments) or keep old hardware around (which is what most electroacoustic studios do these days).
SC decided for a semi-stable rolling release cycle - there are no LTS versions and there is still stuff breaking in new versions, so the worst of both worlds in this regard.

Maybe we should start maintaining a LTS version (which bumps the patch version occasionally), but I don't see available resources for that as we barely keep up with new platform requirements already.

> We should really be working toward SC4. It's time.

I really don't know what SC4 should be like - there are already n libraries/engines out there these days which can to what SC can do, but SC comes with a ton of (legacy) extensions which cover most of ones needs and is probably the reason why it still has at least some relevancy. If we break with that we would loose all the benefits of SC, so I'd be reluctant of doing that.
We had an infamous forum thread, so I don't want to go deeper into that.

The main problem is IMO that people work within an implicit runtime and later expect another implicit runtime to behave the same. By making the runtime explicit for a project it is possible to do "bigger" changes (i.e. this one) while maintaining backwards comparability upon request.
I.e. we should make everything modular (UGens, ClassLibrary, ...) and then a person can specify runtime, class library, UGens, and extensions to use, such that I can run sclang 3.15 with 3.13 class library (the sclang runtime should always be backwards compatible to previous class library) with the UGens of 3.13 (to get a version before e.g. https://github.com/supercollider/supercollider/pull/6833) and version x of a ServerExtension and some Quarks.
SC therefore becomes more like a distribution which can then be adjusted for each project environment - IMO this is the way to go and is also what every other programming language is doing for a good reason.

As mentioned, I am working on such a package manager to have it available for the server-plugin-breaking of 3.15.

edit: regarding dyfers vote - there is the option missing to just replace the implementation as I also don't know a single language who does not truncate to zero. If we have a proper packaging system and actually make use of versioning we could just replace the broken implementation - I am really against keeping wrong stuff around, it just adds overhead for everyone - the users and maintainers.

**Comment by telephon** [comment]
_It would be really nice if we could talk about such major topics in the dev meeting and not at the tail of a contentious topic._ 

Would it be okay to conclude this side-thread by saying that there is no easy escape from such trade-offs and actually being a happy programmer in collective means finding a dimension of joy in them? Just for now?

Then, I'd be really happy to see a pull request for a trunc2 – no other programming language has our trunc, so this is a feature we should not be angry about. Maybe it is a bit comical, I admit, but formally, we are not wrong! 

By the way, just for reference, while James McCartney has changed and improved many sclang things in SAPF, he has kept trunc the same: https://github.com/lfnoise/sapf/blob/main/src/MathOps.cpp. The definition could help for documentation: `DEF2(trunc, "round x to nearest multiple of y <= x")`

**Comment by telephon** [comment]
Just to make sure, this may have come across as too strong: if necessary, please do continue to discuss, independent of my view on the topic, which only matters so much and also may change. I would love it if we could come to a real agreement about this.

**Comment by dyfer** [comment]
At the dev meeting we decided to fix this by:
- documenting current behavior of .trunc
- introducing a new method with desired behavior (truncate towards 0)

We need to decide on the name for the new function. The proposed names are:

🚀 `trunc2` (as in clip2 being symmetrical)
👀 `truncate`

Let's vote on the solution!

**Comment by telephon** [comment]
As a reference:

<img width="512" height="440" alt="Image" src="https://github.com/user-attachments/assets/d1764810-3e92-4254-913f-c1ce2f1c4e47" />

```
(0..100).normalize(-5, 5).collect { |a| [a.trunc, a.abs.trunc * a.sign] }.flop.plot("1. sclang trunc, 2. usual trunc (trunc2)")
```

**Comment by flowerornament** [comment]
Can't wait for SC4. This thread was disheartening to read as a newer user—I wouldn't want to invest my time learning a language that's so committed to backward compatibility they defend illogical design choices.

But then again, perhaps I'm used to norms from the programming world. In any normal workflow, a project is pinned to a specific language version, so this is a non-issue.

**Comment by capital-G** [comment]
> Can't wait for SC4. This thread was disheartening to read as a newer user—I wouldn't want to invest my time learning a language that's so committed to backward compatibility they defend illogical design choices.
> 
> But then again, perhaps I'm used to norms from the programming world. In any normal workflow, a project is pinned to a specific language version, so this is a non-issue.

I feel the pain, and quoting myself which is the fundamental problem as you also noticed

> The main problem is IMO that people work within an implicit runtime and later expect another implicit runtime to behave the same. By making the runtime explicit for a project it is possible to do "bigger" changes (i.e. this one) while maintaining backwards comparability upon request.

Being stuck is probably the class libraries biggest flaw currently - but at least it is reliable. But by being more explicit, we can have both.
I don't think we need SC4 to fix broken implementations, but having a way to not break existing code is always beneficial. In best case we have 3.15 with the new package manager out, which we can use to establish people to use the package manager during 3.15 and can reference 3.15 as fix point for anything w/o major breakage - starting from 3.16 we could start to fix broken behavior as we can simply tell people "tell the package manager that your project is targeting 3.15, which will be also used as a fallback if you rely on old behavior or do not have specified a version in the first place".

We shouldn't be too disruptive though - being settled also has some benefits and I don't want the sclang environment to mirror the npm world.
